sound: Handle Pipewire prepare function

dorindabassey opened the pull request:
Add functions to create, process and connect the stream in the prepare function.

bilelmoussaoui commented on crates/sound/Cargo.toml:
```diff
@@ -25,10 +25,10 @@ virtio-bindings = "0.2"
-pipewire-sys = { version = "0.6.0", optional = true }
-libspa-sys = { version = "0.6.0", optional = true }
+pipewire = { git = "https://gitlab.freedesktop.org/pipewire/pipewire-rs.git", rev = "a144cbd249ce4de34140c3e5e5c70490180cdc05", optional = true }
+libspa = { git = "https://gitlab.freedesktop.org/pipewire/pipewire-rs.git", rev = "a144cbd249ce4de34140c3e5e5c70490180cdc05", optional = true }
```
Unrelated, but you don't need to add libspa/pipewire-sys/libspa-sys dependencies explicitly as they are re-exported by pipewire, see https://gitlab.freedesktop.org/pipewire/pipewire-rs/-/blob/main/pipewire/src/lib.rs#L119-120. So you could do `use pipewire::spa::SomeSpaType;`

dorindabassey replied:
ok I'll try that thanks :)

bilelmoussaoui commented on crates/sound/src/audio_backends/pw_backend.rs:
```diff
@@ -160,13 +202,174 @@ impl AudioBackend for PwBackend {
+            pos[1] = SPA_AUDIO_CHANNEL_FR;
+            let mut info = spa_audio_info_raw {
+                format: match params.format {
+                    VIRTIO_SND_PCM_FMT_S8 => SPA_AUDIO_FORMAT_S8,
```
I feel like those `VIRTIO_SND_PCM_FMT_` constants should be an enum instead and then implement some function to give you the `SPA_AUDIO_FORMAT_` equivalent. Ideally the later would also be a proper enum exposed by the `libspa` bindings from upstream.

dorindabassey replied:
yes, we could use enum. /me thinks about it.

dorindabassey replied:
I agree it's best to put them in the libspa bindings upstream. Thanks :)

bilelmoussaoui replied:
But  `VIRTIO_SND_PCM_FMT_` comes from virtio-sound

dorindabassey replied:
yes it comes from virtio-sound, and other parts of the virtio-sound uses those constants, I will prefer to leave it as constants.

bilelmoussaoui commented on crates/sound/src/audio_backends/pw_backend.rs:
```diff
@@ -160,13 +202,174 @@ impl AudioBackend for PwBackend {
+                    _ => SPA_AUDIO_FORMAT_UNKNOWN,
+                },
+                rate: match params.rate {
+                    VIRTIO_SND_PCM_RATE_32000 => 32000,
```
Same here, should very probably be an enum for refresh rates.

bilelmoussaoui commented on crates/sound/src/audio_backends/pw_backend.rs:
```diff
@@ -160,13 +202,174 @@ impl AudioBackend for PwBackend {
+                properties! {
+                    *pw::keys::MEDIA_TYPE => "Audio",
+                    *pw::keys::MEDIA_CATEGORY => "Playback",
+                    *pw::keys::MEDIA_ROLE => "Music",
```
Is this correct?

dorindabassey replied:
this is just metadata, I'm not pretty sure about the media_role, but I can decide to not set it(not necessary). given that we are only configuring the metadata for playback ATM, the properties metadata should be correct.

bilelmoussaoui commented on crates/sound/src/audio_backends/pw_backend.rs:
```diff
@@ -160,13 +202,174 @@ impl AudioBackend for PwBackend {
+                p
+            };
+
+            let _context = &self.context;
```
```suggestion
```

dorindabassey replied:
this is intentional, because the context doesn't stay alive at the end of the function we need to find a way to keep it alive to be able to use core in the prepare method. see related bug - https://gitlab.freedesktop.org/pipewire/pipewire-rs/-/issues/64

bilelmoussaoui replied:
But in your context is already part of Self so it should not be dropped at the end of the function, if it does, something else is wrong

dorindabassey replied:
Because it is part of Self, if you don't use it you get `field `context` is never read`

bilelmoussaoui replied:
It is still wrong to do this. You should do `#[allow(dead_code)]` before the field and add a comment explaining why.

bilelmoussaoui commented on crates/sound/src/audio_backends/pw_backend.rs:
```diff
@@ -160,13 +202,174 @@ impl AudioBackend for PwBackend {
+            .expect("could not create new stream");
+
+            let thread_clone = self.thread_loop.clone();
+            let _listener_stream = stream
```
```suggestion
            stream
```

dorindabassey replied:
this is intentional because we need to save the listener returned by .register() in a stack variable

bilelmoussaoui commented on crates/sound/src/vhu_sound.rs:
```diff
@@ -336,10 +348,39 @@ impl VhostUserSoundThread {
+                        .read_obj::<VirtioSoundPcmHeader>(desc_request.addr())
+                        .map_err(|_| Error::DescriptorReadFailed)?;
+                    let stream_id = u32::from(pcm_hdr.stream_id);
+                    dbg!("stream_id: {}", stream_id);
```
```suggestion
```

bilelmoussaoui commented on crates/sound/src/vhu_sound.rs:
```diff
@@ -188,7 +188,7 @@ impl VhostUserSoundThread {
         for desc_chain in requests {
             let descriptors: Vec<_> = desc_chain.clone().collect();
-            debug!("Sound request with n descriptors: {}", descriptors.len());
+            //debug!("Sound request with n descriptors: {}", descriptors.len());
```
There are various changes commenting debug statements, not sure if those are meant to be committed

dorindabassey replied:
no no, they're not meant to be committed, this was just for testing. I would clean things up. Thanks :)

bilelmoussaoui commented on crates/sound/src/audio_backends.rs:
```diff
@@ -14,7 +14,12 @@ use crate::PCMParams;
 pub trait AudioBackend {
+    fn get_stream_state(&self, _stream_id: u32) -> Result<u32> {
+        Ok(0)
+    }
```
Does it make sense to provide a default stream state ? every audio backend should be implementing it no?

bilelmoussaoui replied:
Also shouldn't the stream state be some enum instead of a random u32?

MatiasVara replied:
Yes, state shall be an enum and each audio backend shall implement one. How do you provide a default stream state?

dorindabassey replied:
this stream state is generic , because each audio backend should get the stream state from the control queue. cc @MatiasVara for response.

MatiasVara replied:
I think what @bilelmoussaoui means is to removed the generic definition since some backend may rely on it and is wrong.  So that line should only contain the signature.

bilelmoussaoui commented on crates/sound/src/audio_backends/pw_backend.rs:
```diff
@@ -3,21 +3,50 @@
 
 use super::AudioBackend;
 use crate::vhu_sound::NR_STREAMS;
+use crate::virtio_sound::*;
```
this is bad

dorindabassey replied:
yeah, I forgot to do something about that thanks :)

bilelmoussaoui commented on crates/sound/src/audio_backends/pw_backend.rs:
```diff
@@ -3,21 +3,50 @@
+use libspa_sys::SPA_AUDIO_FORMAT_U32;
+use libspa_sys::SPA_AUDIO_FORMAT_U8;
+use libspa_sys::SPA_AUDIO_FORMAT_UNKNOWN;
+use libspa_sys::{spa_audio_info_raw, spa_pod, SPA_AUDIO_FORMAT_S16};
```
All of those imports should be grouped ideally

MatiasVara commented on crates/sound/src/audio_backends/pw_backend.rs:
```diff
@@ -160,13 +202,174 @@ impl AudioBackend for PwBackend {
+
+            stream
+                .connect(
+                    spa::Direction::Output,
```
I think Direction shall be based on parameters of the stream.

MatiasVara commented on crates/sound/src/audio_backends/pw_backend.rs:
```diff
@@ -133,16 +157,26 @@ impl PwBackend {
 }
 
 impl AudioBackend for PwBackend {
+    fn get_stream_state(&self, stream_id: u32) -> Result<u32> {
```
Do we need to check if `stream_id` is a valid id? We should display a msg if the id is invalid I think.

dorindabassey replied:
I don't think it's necessary, and there's already a check for states at vhu_sound.rs:L361

bilelmoussaoui commented on crates/sound/Cargo.toml:
```diff
@@ -25,10 +25,8 @@ virtio-bindings = "0.2"
-pipewire-sys = { version = "0.6.0", optional = true }
-libspa-sys = { version = "0.6.0", optional = true }
+pipewire = { git = "https://gitlab.freedesktop.org/pipewire/pipewire-rs.git", rev = "a144cbd249ce4de34140c3e5e5c70490180cdc05", optional = true }
+pipewire-sys = { git = "https://gitlab.freedesktop.org/pipewire/pipewire-rs.git", rev = "a144cbd249ce4de34140c3e5e5c70490180cdc05", optional = true }
```
```suggestion
```
this one is also not needed, it is re-exported

bilelmoussaoui commented on crates/sound/src/vhu_sound.rs:
```diff
@@ -336,10 +350,38 @@ impl VhostUserSoundThread {
+                        .read_obj::<VirtioSoundPcmHeader>(desc_request.addr())
+                        .map_err(|_| Error::DescriptorReadFailed)?;
+                    let stream_id = u32::from(pcm_hdr.stream_id);
+                    let state = audio_backend.get_stream_state(stream_id).unwrap();
```
```suggestion
                    let state = audio_backend.get_stream_state(stream_id)?;
```

MatiasVara commented on crates/sound/src/audio_backends/pw_backend.rs:
```diff
@@ -210,12 +212,15 @@ impl AudioBackend for PwBackend {
 
-        if stream_states[stream_id as usize] == VIRTIO_SND_R_PCM_SET_PARAMS {
+        if stream_states[stream_id as usize] == VIRTIO_SND_R_PCM_SET_PARAMS
+            || stream_states[stream_id as usize] == VIRTIO_SND_R_PCM_PREPARE
```
I think if guest sends the PREPARE cmd and the stream is in the PREPARE state, this would end up creating the stream again. I think we need to handle the case in which guest sends SET_PARAM and the stream is in the prepare state. In that case, we have to destroy the stream (maybe call release?) and moves the state to SET_PARAM. The next call to PREPARE would create the stream again.

stefano-garzarella commented on crates/sound/src/audio_backends/pw_backend.rs:
```diff
@@ -3,21 +3,43 @@
+use crate::virtio_sound::{VIRTIO_SND_PCM_FMT_U16, VIRTIO_SND_PCM_FMT_U32, VIRTIO_SND_PCM_FMT_U8};
+use crate::virtio_sound::{VIRTIO_SND_PCM_RATE_32000, VIRTIO_SND_PCM_RATE_44100};
+use crate::virtio_sound::{VIRTIO_SND_PCM_RATE_48000, VIRTIO_SND_PCM_RATE_64000};
+use crate::virtio_sound::{VIRTIO_SND_R_PCM_PREPARE, VIRTIO_SND_R_PCM_SET_PARAMS};
```
I would do a single `use crate::virtio_sound::{` with all of them

stefano-garzarella commented on crates/sound/src/audio_backends/pw_backend.rs:
```diff
@@ -160,13 +198,189 @@ impl AudioBackend for PwBackend {
+                            return;
+                        }
+                        let buf = (*b).buffer;
+                        let mut datas = (*buf).datas;
```
```
warning: variable does not need to be mutable
   --> crates/sound/src/audio_backends/pw_backend.rs:306:29
    |
306 |                         let mut datas = (*buf).datas;
    |                             ----^^^^^
    |                             |
    |                             help: remove this `mut`
    |
```

stefano-garzarella commented on crates/sound/src/audio_backends/pw_backend.rs:
```diff
@@ -3,21 +3,43 @@
+use pw::spa::sys::{spa_ringbuffer_get_read_index, spa_ringbuffer_read_update};
+use pw::spa::sys::{SPA_AUDIO_CHANNEL_FL, SPA_AUDIO_CHANNEL_FR, SPA_AUDIO_FORMAT_S32};
+use pw::spa::sys::{SPA_AUDIO_FORMAT_S16, SPA_AUDIO_FORMAT_U8, SPA_AUDIO_FORMAT_UNKNOWN};
+use pw::spa::sys::{SPA_AUDIO_FORMAT_S8, SPA_AUDIO_FORMAT_U16, SPA_AUDIO_FORMAT_U32};
```
Same here

stefano-garzarella commented on crates/sound/src/audio_backends/pw_backend.rs:
```diff
@@ -133,16 +160,27 @@ impl PwBackend {
+    fn get_stream_state(&self, stream_id: u32) -> Result<u32> {
+        let stream_states = self.stream_states.read().unwrap();
+        dbg!("stream states: {}", stream_states.clone());
+        Ok(stream_states[stream_id as usize])
```
If a malicious guest sends a wrong stream_id, could this generate an out-of-boundary panic?

dorindabassey replied:
maybe it might make sense to propagate the error here

Commit:
sound: Handle Pipewire prepare function

Add functions to create, process and connect the stream in the prepare function.
keep stream listener alive and add param changed callback to update params when new set of params
are negotiated.

Co-authored-by: Dorinda Bassey <dbassey@redhat.com>
Co-authored-by: Matias Ezequiel Vara Larsen <mvaralar@redhat.com>

Signed-off-by: Dorinda Bassey <dbassey@redhat.com>
